### PR TITLE
feat: support qwen3-32b/qwen3-30b-a3b precision and performance on mlu backend.

### PR DESCRIPTION
**Pull request by a120092009**
Add MLU Support

### CODE REVIEW

File: xllm/core/distributed_runtime/worker_service.cpp [hunk 1]
```diff
@@ -356,7 +363,7 @@ void WorkerService::ExecuteModel(
 #if defined(USE_NPU)
     c10_npu::SetDevice(device_.index());
 #elif defined(USE_MLU)
-    // TODO(mlu): implement mlu execute model
+    const torch_mlu::mlu::MLUGuard device_guard(device_.index());
```
Comment (XuZhang99) [inline xllm/core/distributed_runtime/worker_service.cpp]: `set_device` is `device_guard` on MLU?

File: xllm/core/framework/eplb/CMakeLists.txt [hunk 2]
```diff
@@ -25,7 +25,7 @@ cc_library(
     shared_memory_manager.cpp
     expert_buffer_manager.cpp
   DEPS
-    llm_engine
+    $<$<BOOL:${USE_NPU}>:llm_engine>
```
Comment (XuZhang99) [inline xllm/core/framework/eplb/CMakeLists.txt]: why add `USE_NPU` for llm_engine?

File: xllm/core/framework/block/CMakeLists.txt [hunk 3]
```diff
@@ -47,5 +47,5 @@ cc_test(
     GTest::gtest_main
 )
 
-target_link_libraries(block_test PRIVATE brpc OpenSSL::SSL OpenSSL::Crypto ascendcl Folly::folly)
+target_link_libraries(block_test PRIVATE brpc OpenSSL::SSL OpenSSL::Crypto $<$<BOOL:${USE_NPU}>:ascendcl> Folly::folly)
```
Comment (XuZhang99) [inline xllm/core/framework/block/CMakeLists.txt]: remove `ascendcl` directly

File: xllm/core/util/threadpool.cpp [hunk 4]
```diff
@@ -17,7 +17,9 @@ limitations under the License.
 
 #include <thread>
 
+#if defined(USE_NPU)
 #include "concurrent_queue.h"
```
Comment (XuZhang99) [inline xllm/core/util/threadpool.cpp]: remove this directly

File: xllm/core/util/type_traits.h [hunk 5]
```diff
@@ -38,10 +38,14 @@ struct remove_optional<std::optional<value_type>> {
...
 
-#if defined(USE_NPU)
 using VirPtr = void*;
+
+#if defined(USE_NPU)
 using PhyMemHandle = aclrtDrvMemHandle;
 using VmmResult = aclError;
+#elif defined(USE_MLU)
+using PhyMemHandle = void*;
+using VmmResult = int;
```
Comment (XuZhang99) [inline xllm/core/util/type_traits.h:48]:
```
#elif defined(USE_MLU)
using VirPtr = CNaddr;
using PhyMemHandle = CNmemGenericAllocationHandle;
using VmmResult = int;
```

Comment (XuZhang99) [reply xllm/core/util/type_traits.h:48]: by the way, complete virtual memory API in xtensor and can remove lots of macro.
you can find the vmm API for mlu here: https://github.com/Cambricon/vllm-mlu/blob/v0.9.1-dev/csrc/cnmem_allocator.cpp

File: xllm/core/framework/prefix_cache/CMakeLists.txt [hunk 6]
```diff
@@ -41,7 +41,7 @@ cc_test(
     GTest::gtest_main
 )
 
-target_link_libraries(prefix_test PRIVATE brpc OpenSSL::SSL OpenSSL::Crypto ascendcl Folly::folly)
+target_link_libraries(prefix_test PRIVATE brpc OpenSSL::SSL OpenSSL::Crypto $<$<BOOL:${USE_NPU}>:ascendcl> Folly::folly)
```
Comment (XuZhang99) [inline xllm/core/framework/prefix_cache/CMakeLists.txt]: why prefix_cache need `ascendcl`?

File: xllm/core/runtime/llm_worker_impl.cpp [hunk 7]
```diff
@@ -28,6 +28,8 @@ limitations under the License.
 #include <torch_npu/torch_npu.h>
 
 #include "pytorch/adapter/utils/utils.h"
+#elif defined(USE_MLU)
+#include <torch_mlu/csrc/framework/core/MLUStream.h>
```
Comment (XuZhang99) [inline xllm/core/runtime/llm_worker_impl.cpp]: why include stream here?

File: xllm/core/scheduler/CMakeLists.txt [hunk 8]
```diff
@@ -36,25 +36,27 @@ cc_library(
...
+      chunked_prefill_scheduler_test
+      continuous_scheduler_test
+    SRCS
+      chunked_prefill_scheduler_test.cpp
+      continuous_scheduler_test.cpp
+    DEPS
+      :scheduler
+      GTest::gtest_main
+  )
+  target_link_libraries(chunked_prefill_scheduler_test
```
Comment (XuZhang99) [inline xllm/core/scheduler/CMakeLists.txt]: remove `target_link_libraries` and add `    $<$<BOOL:${USE_NPU}>:nnopbase>` to `DEPS` for cc_test. `if(USE_NPU)` should be removed.

File: xllm/models/llm/llm_model_base.h [hunk 9]
```diff
@@ -27,13 +29,19 @@ limitations under the License.
...
 #include "core/framework/model/model_input_params.h"
 #include "core/framework/model_context.h"
+#if defined(USE_NPU)
 #include "core/layers/attention_mask.h"
 #include "core/layers/block_copy.h"
 #include "core/layers/lm_head.h"
 #include "core/layers/pos_embedding.h"
 #include "core/layers/rms_norm.h"
 #include "models/model_registry.h"
 #include "xllm_kernels/core/include/atb_speed/log.h"
```
Comment (XuZhang99) [inline xllm/models/llm/llm_model_base.h:40]: only this line should put in `#if defined(USE_NPU)`

File: xllm/models/models.h [hunk 10]
```diff
@@ -31,3 +31,9 @@ limitations under the License.
 #include "vlm/minicpmv.h"         // IWYU pragma: keep
 #include "vlm/qwen2_5_vl.h"       // IWYU pragma: keep
 #endif
+
+#if defined(USE_MLU)
```
Comment (XuZhang99) [inline xllm/models/models.h]: if a model is supported for both npu and mlu, it should not be placed in macro.

File: xllm/core/distributed_runtime/worker_server.cpp [hunk 11]
```diff
@@ -64,7 +66,7 @@ void WorkerServer::create_server(const runtime::Options& options,
     LOG(ERROR) << "ACL set device id: " << device_id << " failed, ret:" << ret;
   }
 #elif defined(USE_MLU)
-// TODO(mlu): implement mlu device set
+  const torch_mlu::mlu::MLUGuard device_guard(device_id);
```
Comment (yq33victor) [inline xllm/core/distributed_runtime/worker_server.cpp]: `guard device conetxt` will be invalid outside this function ?

File: xllm/core/framework/eplb/CMakeLists.txt (hunk 2 above)
Comment (a120092009) [reply xllm/core/framework/eplb/CMakeLists.txt]: error will occur if deleted
1b2d905cd4386441cf337e1feca56771

File: xllm/core/framework/block/CMakeLists.txt (hunk 3 above)
Comment (a120092009) [reply xllm/core/framework/block/CMakeLists.txt]: fixed

File: xllm/core/util/threadpool.cpp (hunk 4 above)
Comment (a120092009) [reply xllm/core/util/threadpool.cpp]: fixed

File: xllm/core/framework/prefix_cache/CMakeLists.txt (hunk 6 above)
Comment (a120092009) [reply xllm/core/framework/prefix_cache/CMakeLists.txt]: not mlu related, you can check it on npu

File: xllm/core/runtime/llm_worker_impl.cpp (hunk 7 above)
Comment (a120092009) [reply xllm/core/runtime/llm_worker_impl.cpp]: fixed

File: xllm/core/scheduler/CMakeLists.txt (hunk 8 above)
Comment (a120092009) [reply xllm/core/scheduler/CMakeLists.txt]: fixed

File: xllm/models/models.h (hunk 10 above)
Comment (a120092009) [reply xllm/models/models.h]: fixed

File: xllm/core/util/type_traits.h (hunk 5 above)
Comment (a120092009) [reply xllm/core/util/type_traits.h:48]: The fist comment 
using VirPtr = CNaddr;
using PhyMemHandle = CNmemGenericAllocationHandle;
using VmmResult = int;

Why do you want to use in this way? Actually we use torch api, all device operation should be wrapped by torch api

File: xllm/core/distributed_runtime/worker_service.cpp (hunk 1 above)
Comment (a120092009) [reply xllm/core/distributed_runtime/worker_service.cpp]: For now, torch_mlu don't expose SetDevice api, we can only use MLUGuard to set device. This problem will be fixed in later version

File: xllm/core/framework/kv_cache/kv_cache.h [hunk 12]
```diff
@@ -21,28 +21,38 @@ limitations under the License.
...
+#if defined(USE_NPU)
 #include "framework/xtensor/xtensor.h"
+#endif
 
 namespace xllm {
 class KVCache final {
  public:
   KVCache() = default;
   KVCache(torch::Tensor key_cache, torch::Tensor value_cache);
+#if defined(USE_NPU)
```
Comment (yq33victor) [inline xllm/core/framework/kv_cache/kv_cache.h]: xtensor is device-agnostic, so theoretically, there's no need to use `if-define` for restrictions?

File: xllm/models/llm/llm_model_base.h (hunk 9 above)
Comment (a120092009) [reply xllm/models/llm/llm_model_base.h:40]: fixed

File: xllm/core/framework/dit_model_context.cpp [hunk 13]
```diff
@@ -72,6 +72,7 @@ const QuantArgs& DiTModelContext::get_quant_args(
   }
 }
 
+#if defined(USE_NPU)
```
Comment (yq33victor) [inline xllm/core/framework/dit_model_context.cpp:75]: nit: maybe we can keep the interface, it’s just that the MLU doesn’t need to call it.

File: xllm/core/framework/eplb/CMakeLists.txt (hunk 2 above)
Comment (XuZhang99) [reply xllm/core/framework/eplb/CMakeLists.txt]: remove `llm_engine` because it's useless

File: xllm/core/runtime/llm_master.cpp [hunk 14]
```diff
@@ -120,7 +122,7 @@ LLMMaster::~LLMMaster() {
 #if defined(USE_NPU)
   c10_npu::NPUCachingAllocator::emptyCache();
 #elif defined(USE_MLU)
-  // TODO(mlu): implement mlu empty cache
+  torch_mlu::MLUCachingAllocator::emptyCache();
```
Comment (XuZhang99) [inline xllm/core/runtime/llm_master.cpp]: all `emptyCache()` can be removed here

File: xllm/core/runtime/llm_master.cpp [hunk 15]
```diff
@@ -38,6 +38,8 @@ limitations under the License.
 #include "server/xllm_server_registry.h"
 #if defined(USE_NPU)
 #include "torch_npu/csrc/core/npu/NPUCachingAllocator.h"
+#elif defined(USE_MLU)
+#include <torch_mlu/csrc/framework/core/caching_allocator.h>
```
Comment (XuZhang99) [inline xllm/core/runtime/llm_master.cpp]:
```
#if defined(USE_NPU)
#include "torch_npu/csrc/core/npu/NPUCachingAllocator.h"
#elif defined(USE_MLU)
#include <torch_mlu/csrc/framework/core/caching_allocator.h>
```
these lines should be removed too.

File: xllm/core/runtime/vlm_master.cpp [hunk 16]
```diff
@@ -125,7 +127,7 @@ VLMMaster::~VLMMaster() {
 #if defined(USE_NPU)
   c10_npu::NPUCachingAllocator::emptyCache();
 #elif defined(USE_MLU)
-  // TODO(mlu): implement mlu empty cache
+  torch_mlu::MLUCachingAllocator::emptyCache();
```
Comment (XuZhang99) [inline xllm/core/runtime/vlm_master.cpp]: same as above

File: xllm/core/framework/parallel_state.h [hunk 17]
```diff
@@ -45,9 +48,17 @@ torch::Tensor reduce(torch::Tensor input, const ParallelArgs& parallel_args);
 
 torch::Tensor scatter(torch::Tensor input, const ParallelArgs& parallel_args);
 
+#if defined(USE_MLU)
+torch::Tensor gather(torch::Tensor input, ProcessGroup* process_group);
```
Comment (yq33victor) [inline xllm/core/framework/parallel_state.h:45]: could we use the defination above `torch::Tensor gather(torch::Tensor input, const ParallelArgs& parallel_args);`?  Because we can get `process_group` from `parallel_args`.

File: xllm/core/framework/prefix_cache/CMakeLists.txt (hunk 6 above)
Comment (XuZhang99) [reply xllm/core/framework/prefix_cache/CMakeLists.txt]: remove `ascendcl`

File: xllm/core/platform/device.cpp [hunk 18]
```diff
@@ -28,15 +34,15 @@ void Device::init_device_context() const {
 #if defined(USE_NPU)
   torch_npu::init_npu(device_name);
 #elif defined(USE_MLU)
-  // TODO(mlu): implement mlu init context
+  const torch_mlu::mlu::MLUGuard device_guard(index());
```
Comment (yq33victor) [inline xllm/core/platform/device.cpp]: `guard` , the same question as above.

File: xllm/models/llm/llm_model_base.h [hunk 19]
```diff
@@ -15,7 +15,9 @@ limitations under the License.
 
 #pragma once
 
+#if defined(USE_NPU)
 #include <atb/atb_infer.h>
+#endif
```
Comment (yq33victor) [inline xllm/models/llm/llm_model_base.h:20]: ok, here, could we create new classes for MLU.
 emmm, there are too many `if-define`  conditional judgment branch statements. :)

File: xllm/models/llm/qwen3.h [hunk 20]
```diff
@@ -46,6 +46,9 @@ class QWen3ModelImpl : public QWenModelImplBase<QWen3DecoderLayer> {
 
     blocks_ = register_module("layers", torch::nn::ModuleList());
     layers_.reserve(model_args.n_layers());
+#if defined(USE_NPU)
+    embed_tokens_ =
```
Comment (yq33victor) [inline xllm/models/llm/qwen3.h]: nit: delete this line code,  npu use the code below:`embed_tokens_.push_back(layer::WordEmbedding(context));`

File: xllm/core/runtime/llm_master.cpp (hunk 14 above)
Comment (a120092009) [reply xllm/core/runtime/llm_master.cpp]: why does npu need empty_cache, but mlu not?

Comment (XuZhang99) [reply xllm/core/runtime/llm_master.cpp]: both npu and mlu should be removed

File: xllm/core/layers/mlu/qwen3_decoder_layer.h [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+
+  torch::Tensor forward(torch::Tensor& x,
+                        torch::Tensor& positions,
+                        torch::Tensor& residual,
+                        const AttentionMetadata& attn_metadata,
+                        KVCache& kv_cache,
+                        const ModelInputParams& input_params);
+
+ private:
+  Qwen3Attention attention_;
```
Comment (XuZhang99) [inline xllm/core/layers/mlu/qwen3_decoder_layer.h]: why not use ptr for `attention_`

File: xllm/core/layers/mlu/qwen3_decoder_layer.h [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+  Qwen3MLP mlp_{nullptr};
+  RmsNorm input_norm_{nullptr};
+  RmsNorm post_norm_{nullptr};
+
+  c10::ScalarType dtype_;
```
Comment (a120092009) [inline xllm/core/layers/mlu/qwen3_decoder_layer.h]: 用torch api

File: xllm/core/distributed_runtime/worker_service.cpp (hunk 1 above)
Comment (a120092009) [reply xllm/core/distributed_runtime/worker_service.cpp]: delete mluguard

File: xllm/core/layers/mlu/qwen3_decoder_layer.h (hunk 21 above)
Comment (a120092009) [reply xllm/core/layers/mlu/qwen3_decoder_layer.h]: fixed

File: xllm/core/platform/device.cpp (hunk 18 above)
Comment (a120092009) [reply xllm/core/platform/device.cpp]: fixed

File: xllm/core/framework/parallel_state.h (hunk 17 above)
Comment (a120092009) [reply xllm/core/framework/parallel_state.h:45]: fixed, only keep ProcessGroup as input argument

File: xllm/core/distributed_runtime/worker_server.cpp (hunk 11 above)
Comment (a120092009) [reply xllm/core/distributed_runtime/worker_server.cpp]: deleted

**Review by liutongxuan — APPROVED** [review]
